design/secure-communication.md

```diff
@@ -0,0 +1,94 @@
+
+## Detailed Design - mTLS
+
+In order for mTLS to work, each end must know:
+- The CA that the TLS relationship will descend from.
+- Its certificate
+- Its key
+
+This applies both to Contour and Envoy.
+
+To accomplish this, we will add the following new command line options:
+
+To `contour serve`, covering the Contour side:
+- `--contour-cafile`: the CA Bundle for the mutual TLS relationship
+- `--contour-mtls-cert`: The cert for Contour to use to identify itself. (`serve`)
+- `--contour-mtls-key`: The key for Contour to use for its side of the mTLS pairing. (`serve`)
+
+And to `contour bootstrap`, covering Envoy configuration:
+- `--envoy-cafile`: The CA Bundle for the mutual TLS relationship
+- `--envoy-mtls-cert`: The cert for Envoy to use to identify itself.
+- `--envoy-mtls-key`: The key for Envoy to use to identify itself.
+
+If any one of the three options is passed to either `serve` or `bootstrap`, the other two are required, and we will error out if they are not passed.
+
+We will wire these options into the gRPC server establishment process on the Contour side, and supply them in the bootstrap on the Envoy side. This is where the most significant code changes will be required.
+
+We will update the example deployment manifests for the separate deployment options to use these new options, and provide a basic sample howto on generating a CA keypair and mTLS keypairs using command line tools like openssl.
+
+## Security Considerations
+
+In the current design, these certs require the rolling of the deployment to change, they should be long-lived certs (days, weeks, or months), not short-lived (minutes or hours). The CA Keypair should be very long-lived and very tightly controlled, as Contour/Envoy connection security is only as secure as the CA keypair.
+
+As noted above, this is an initial design to *start* the process of being able to run reasonably secure, authenticated, separate deployments of Contour and Envoy. Some things that we *will* address at a later date, but are out of scope for now:
+
+- Certificate generation (apart from a very simple howto to allow users to see how the feature works)
+- Certificate rotation
+
+Implementing this change will, however, improve the security of the split-deployment scenario.
+
+## Future Work
+
+This work is the first phase of a planned three-phase rollout.
+
+### Phase 1: mTLS
+
+This document - allows a more secure split deployment in a very basic way.
+
+### Phase 2: User Experience
+
+This phase will add additional functionality and/or tools to help with the certificate generation and rolling process. Watch [#1184][1] to track this phase.
+
+### Phase 3: Secure by default
+
+This phase will flip the example deployments to all use a secure mode, make those command line options default, and include an option to restore the current behaviour. Watch [#1185][2] to track this phase.
```

Comment:
Build error with spelling: `"behaviour" is a misspelling of "behavior"`

Reply:
American vs UK spelling FTL. I'll change it to the US spelling.

Reply:
Sorry. I lost that battle early on, but rest assured I am doing my best to sneak other aspects of the monarchy into this project.